### PR TITLE
feat: add first-class PostgreSQL persistence support

### LINKED ISSUE #107: Support PostgreSQL alongside MySQL with a pluggable database access layer
## Background

Many international self-hosted users standardize on PostgreSQL for operational databases. kkrepo currently documents and implements MySQL as the shared metadata, security, token, audit, migration, session, and coordination store. To make deployment easier in PostgreSQL-first environments, kkrepo should allow operators to choose either PostgreSQL or MySQL without changing product behavior.

## Goal

Introduce a database access abstraction that lets kkrepo run on either PostgreSQL or MySQL with the same repository, security, session, migration, and audit semantics.

## Scope

- Keep MySQL as a supported database backend.
- Add PostgreSQL as a first-class supported database backend.
- Abstract database-specific SQL, JSON handling, enum/value mapping, pagination, upsert behavior, locking, and timestamp behavior behind clear interfaces or dialect-specific modules.
- Provide database-specific Flyway migrations or a compatible migration strategy for both MySQL and PostgreSQL.
- Make the active database backend selectable through configuration, Docker Compose, Helm/Kubernetes values, and documentation.
- Ensure runtime behavior remains compatible across database backends, especially for permissions, tokens, audit logs, migration state, repository metadata, sessions, cache versions, and distributed coordination state.

## Acceptance criteria

- kkrepo starts successfully with either MySQL or PostgreSQL using documented configuration.
- Existing DAO/service tests run against both database backends, or an equivalent CI matrix covers both.
- Database-specific behavior is isolated from business logic and protocol handlers.
- The quickstart/deployment documentation explains how to choose MySQL or PostgreSQL.
- Existing MySQL installations keep working without migration surprises.
- PostgreSQL deployments support the same multi-replica correctness guarantees as MySQL-backed deployments.

## Notes

This should be treated as a persistence-layer product capability, not only a driver swap. The implementation needs to preserve transactional uniqueness, distributed session storage, short-lived coordination state, cache version watermarks, migration resume state, and audit visibility across both supported databases.

### PR DESCRIPTION
**Pull request by klboke**
## Summary

- Complete PR 3 through PR 5 of the pluggable database access layer design on top of the existing PR 1/PR 2 boundary and dialect work.
- Add first-class PostgreSQL 12+ persistence with a PostgreSQL dialect, a logical V29 baseline, JSONB/search/upsert/coordination behavior, explicit database selection, and JDBC metadata validation before Flyway migration.
- Preserve MySQL as the default backend, move V1-V29 byte-for-byte into the MySQL backend, protect the legacy checksums, and enforce Flyway version parity for future migrations.
- Make the shared JDBC implementation portable and keep database-specific SQL behind backend SPI implementations.
- Add real MySQL/PostgreSQL contract tests with PostgreSQL 12 as the minimum-version gate and PostgreSQL 16 in end-to-end compatibility, two-instance server smoke tests, concurrency coverage, stable CI matrix checks, PostgreSQL quickstart/dev/compat Compose support, and a multi-replica Helm chart.
- Update the English and Chinese architecture, deployment, backup, troubleshooting, security, migration, compatibility, and database schema documentation.

## Why

The previous runtime and migration layout assumed MySQL even after the public persistence boundary and MySQL dialect had been extracted. PostgreSQL support requires equivalent schema, transaction, JSON, search, session, audit, token, migration, cache-watermark, worker-claim, and upload-session behavior; adding only a JDBC driver would leave correctness gaps in multi-replica deployments.

## User and developer impact

- Existing installations continue to use MySQL by default and retain the original V1-V29 migration contents and baseline behavior.
- New installations can select PostgreSQL with `KKREPO_DATABASE_TYPE=postgresql` and the PostgreSQL JDBC URL/credentials.
- One executable server jar contains both database drivers, both Flyway database modules, and both persistence backends.
- Protocol, controller, business service, and Nexus migration code remain isolated from concrete JDBC backend packages.
- Operators now have PostgreSQL quickstart, Compose, Helm, backup/restore, and troubleshooting guidance.

## Validation

- `mvn -pl server -am -Dsurefire.failIfNoSpecifiedTests=false clean verify`
  - 24 reactor modules passed.
  - Server: 951 tests passed.
  - MySQL persistence: 45 tests passed.
  - PostgreSQL persistence: 18 tests passed.
  - The server smoke matrix starts two full application instances per backend and verifies fresh/repeat migration, health, administrator and repository visibility, cross-instance artifact blobs, Spring Session, audit, cache versions, and marker coordination.
- `mvn -pl compat-test -am -Dtest=MavenMetadataMergeCompatibilityTest,MavenWritePolicyCompatibilityTest,NpmProtocolCompatibilityTest -Dsurefire.failIfNoSpecifiedTests=false verify` (16 tests passed).
- `mvn -pl server -am -DskipTests package spring-boot:repackage`, followed by executable-jar inspection for both JDBC drivers, Flyway modules, backend jars, and migration resources.
- `mvn -pl persistence-postgresql -am -Dtest=PostgreSqlPersistenceApiContractTest,PostgreSqlMigrationCompatibilityTest -Dsurefire.failIfNoSpecifiedTests=false test` against PostgreSQL 12.22 (16 tests passed).
- `mvn -pl server -am -Dtest=DatabaseServerSmokeTest#postgresqlSupportsFreshAndRepeatStartupWithCrossNodeSession -Dsurefire.failIfNoSpecifiedTests=false test` against PostgreSQL 12.22.
- `CLIENT_E2E_TESTS=pypi scripts/ci/run-client-e2e.sh` against the PostgreSQL compatibility stack (build, upload, group simple-index resolution, and pip install passed).
- `scripts/ci/check-flyway-parity.sh`.
- Docker Compose config validation for MySQL/PostgreSQL quickstart, dev, and compatibility profiles.
- `helm lint deploy/helm/kkrepo`.
- GitHub workflow YAML parsing and `git diff --check`.

JaCoCo results after the full verification:

- PostgreSQL backend line coverage: 136/140 (97.14%).
- PostgreSQL backend branch coverage: 29/40 (72.5%).
- Aggregate line coverage: 24,901/34,612 (71.94%).

Closes #107

### CODE REVIEW

**Comment by klboke** [comment]
@codex review

File: persistence-jdbc/src/main/java/com/github/klboke/kkrepo/persistence/jdbc/internal/JdbcProxyStateDao.java [hunk 1]
```diff
@@ -57,35 +58,27 @@ public boolean isBlocked(long repositoryId, Instant now) {
...
+      return;
+    }
+    try {
+      jdbcTemplate.update("""
+          INSERT INTO proxy_remote_state
+            (repository_id, blocked_until, fail_count, last_success_at, last_failure_at, last_error)
+          VALUES (?, NULL, 0, ?, NULL, NULL)
+          """, repositoryId, recordedAt);
+    } catch (DuplicateKeyException concurrentInsertOrThrottledExistingRow) {
+      // A healthy row inside the throttle window is deliberately a no-op.
```
[the remark this answers is not in the document]
Comment (klboke) [reply persistence-jdbc/src/main/java/com/github/klboke/kkrepo/persistence/jdbc/internal/JdbcProxyStateDao.java]: Fixed in `a2bf935`. The `DuplicateKeyException` branch now retries the same conditional clear UPDATE, so a concurrent failure insert is immediately cleared while an already-healthy throttled row remains a no-op. Added a deterministic regression test that verifies the `clear -> insert race -> clear retry` sequence.

File: persistence-postgresql/src/main/resources/db/migration/postgresql/V29__postgresql_baseline.sql [hunk 2]
```diff
@@ -0,0 +1,952 @@
...
+  repository_id BIGINT NOT NULL CHECK (repository_id >= 0),
+  parent_id BIGINT CHECK (parent_id >= 0),
+  component_id BIGINT CHECK (component_id >= 0),
+  asset_id BIGINT CHECK (asset_id >= 0),
+  path VARCHAR(2048) NOT NULL,
+  path_hash BYTEA NOT NULL,
+  display_name VARCHAR(512) NOT NULL,
+  package_url VARCHAR(2048),
+  depth INTEGER DEFAULT 0 NOT NULL,
+  has_asset_subtree BOOLEAN DEFAULT false NOT NULL,
```
[the remark this answers is not in the document]
Comment (klboke) [reply persistence-postgresql/src/main/resources/db/migration/postgresql/V29__postgresql_baseline.sql:126]: Fixed in `a2bf935`. Browse-node subtree reactivation now binds real boolean values, and ancestor pruning uses a boolean predicate instead of comparing the PostgreSQL `BOOLEAN` column with `1`. I also added the cross-backend `browseSubtreeFlagsSupportBooleanReactivationAndPruning` contract test; it passes against MySQL 8 and PostgreSQL 12, along with both server-smoke jobs.

**Comment by klboke** [comment]
@codex review

File: persistence-jdbc/src/main/java/com/github/klboke/kkrepo/persistence/jdbc/internal/support/JdbcUpserts.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      Object[] updateArguments,
+      String insertSql,
+      Object[] insertArguments) {
+    if (jdbc.update(updateSql, updateArguments) > 0) {
+      return;
+    }
+    try {
+      jdbc.update(insertSql, insertArguments);
+    } catch (DuplicateKeyException raceOrDifferentConstraint) {
+      if (jdbc.update(updateSql, updateArguments) == 0) {
```
[the remark this answers is not in the document]
Comment (klboke) [reply persistence-jdbc/src/main/java/com/github/klboke/kkrepo/persistence/jdbc/internal/support/JdbcUpserts.java]: Fixed in `1d3f5fe`. `JdbcUpserts` now uses a JDBC savepoint only for PostgreSQL transactions, rolls back to it before classifying a duplicate key, and retries the UPDATE on the same connection. MySQL retains its portable no-savepoint retry path because its driver can invalidate these savepoints during this upsert flow. Added deterministic real-backend coverage for both sides: PostgreSQL 12 forces two outer transactions through a duplicate insert race and proves the losing transaction recovers and commits; MySQL 8 rejects any savepoint creation while verifying the transactional insert succeeds. Both persistence contracts, both server-smoke jobs, client E2E, all migration matrices, CodeQL, Codecov, and OCI conformance pass.

**Comment by klboke** [comment]
@codex review

**Comment by klboke** [comment]
@codex review

File: persistence-jdbc/src/main/java/com/github/klboke/kkrepo/persistence/jdbc/internal/JdbcProxyStateDao.java [hunk 4]
```diff
@@ -51,41 +53,40 @@ public boolean isBlocked(long repositoryId, Instant now) {
...
+    int updated = clearFailureState(repositoryId, recordedAt, throttleBoundary);
+    if (updated > 0) {
+      return;
+    }
+    try {
+      jdbcTemplate.update("""
+          INSERT INTO proxy_remote_state
+            (repository_id, blocked_until, fail_count, last_success_at, last_failure_at, last_error)
+          VALUES (?, NULL, 0, ?, NULL, NULL)
+          """, repositoryId, recordedAt);
```
[the remark this answers is not in the document]
Comment (klboke) [reply persistence-jdbc/src/main/java/com/github/klboke/kkrepo/persistence/jdbc/internal/JdbcProxyStateDao.java:81]: Fixed in `f87c41c`. `recordSuccess` now reads the shared state first and immediately returns for an already-healthy row inside the throttle window, so the hot path is one indexed read with no write and no duplicate-key exception. Loaded failure/stale-success rows are still cleared immediately; the duplicate-key catch is now reserved for a genuine missing-row insert race and still retries the conditional clear so success wins against a concurrent failure insert. Added deterministic tests for the throttled healthy path, loaded failure recovery, and the missing-row concurrent insert race. All CI checks, CodeQL, Codecov, both database contracts/smokes, client E2E, OCI conformance, and all migration matrices pass.

**Comment by klboke** [comment]
@codex review

File: persistence-postgresql/src/main/resources/db/migration/postgresql/V29__postgresql_baseline.sql [hunk 5]
```diff
@@ -0,0 +1,952 @@
...
+  owner_source VARCHAR(100) NOT NULL,
+  owner_user_id VARCHAR(255) NOT NULL,
+  display_name VARCHAR(255),
+  status VARCHAR(50) DEFAULT 'ACTIVE' NOT NULL,
+  api_key_hash VARCHAR(255) NOT NULL,
+  token_prefix VARCHAR(32),
+  scopes_json JSONB,
+  encrypted_payload TEXT NOT NULL,
+  expires_at TIMESTAMPTZ(3),
+  last_used_at TIMESTAMPTZ(3),
```
[the remark this answers is not in the document]
Comment (klboke) [reply persistence-postgresql/src/main/resources/db/migration/postgresql/V29__postgresql_baseline.sql]: Fixed in `526b2e9`. PostgreSQL now maps all `LocalDateTime`-backed API-key columns (`created_at`, `updated_at`, `expires_at`, and `last_used_at`) to `timestamp without time zone`, preserving the existing MySQL `DATETIME(3)` wall-clock semantics. I applied the same mapping to `security_audit_log.occurred_at`, whose shared model is also `LocalDateTime`. Added a real PostgreSQL 12 regression test that sets the database session timezone to differ from the JVM, asserts exact millisecond round trips for API-key expiry/last-used and audit timestamps, and verifies the column types via `information_schema`. Updated the English and Chinese docs to distinguish instant timestamps from wall-clock `LocalDateTime` fields. All CI, CodeQL, Codecov, database contracts/smokes, client E2E, OCI, and every migration matrix job pass.

**Comment by klboke** [comment]
@codex review
